### PR TITLE
Aave tables layer3

### PR DESCRIPTION
**Pull request by Treaqura**
Created AAVE abstractions for borrowing, lending, interest & token list. The abstractions are on Ethereum, Polygon & Optimism and I included a Dune query to test if they worked.

Ethereum testing:
1. Aave borrow:   https://dune.com/queries/875753
2. Aave deposit:  https://dune.com/queries/875769
3. Aave interest: https://dune.com/queries/875574
4. Aave tokens:   https://dune.com/queries/875561

Optimism testing:
1. Aave borrow:   https://dune.com/queries/875461
2. Aave deposit:  https://dune.com/queries/875485
3. Aave interest: https://dune.com/queries/875509
4. Aave tokens:   https://dune.com/queries/875536

Polygon testing:
1. Aave borrow:   https://dune.com/queries/875384
2. Aave deposit:  https://dune.com/queries/875396
3. Aave interest: https://dune.com/queries/875423
4. Aave tokens:   https://dune.com/queries/875437

* [ ] the query produces the intended results
* [ ] the folder name matches the schema name
* [ ] the schema name exists in Dune
* [ ] views are prefixed with `view_`, functions with `fn_`.
* [ ] the filename matches the defined view, table or function and ends with .sql
* [ ] each file has only one view, table or function defined  
* [ ] column names are `lowercase_snake_cased`

### CODE REVIEW

File: ethereum/aave/borrow.sql [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    
+FROM (
+-- v1
+-- borrow
+SELECT 
+    '1' AS version,
+    'borrow' AS transaction_type,
+    CASE 
+    WHEN "_borrowRateMode" = '1' THEN 'stable'
+    WHEN "_borrowRateMode" = '2' THEN 'variable'
```
Comment (chuxinh) [inline ethereum/aave/borrow.sql:65]: indent

File: optimism2/aave/deposit.sql [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    evt_block_number
+FROM aave_v3."Pool_evt_Withdraw"
+UNION ALL 
+-- liquidation
+SELECT 
+    '3' AS version,
+    'deposit_liquidation' AS transaction_type,
+    "collateralAsset" AS contract_address,
+    "user" AS borrower,
+    liquidator AS "to",
```
Comment (chuxinh) [inline optimism2/aave/deposit.sql:91]: rename them accordingly too

File: optimism2/aave/tokens.sql [hunk 3]
```diff
@@ -0,0 +1,67 @@
+CREATE TABLE IF NOT EXISTS aave.tokens (   
+    symbol text,
+    contract_address bytea PRIMARY KEY,
+    "aToken" bytea,
+    "stableDebtToken" bytea,
+    "variableDebtToken" bytea,
```
Comment (chuxinh) [inline optimism2/aave/tokens.sql:6]: update these columns to `lowercase_snake_cased`

File: optimism2/aave/tokens.sql [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+);
+
+CREATE OR REPLACE FUNCTION aave.insert_tokens(start_ts timestamptz, end_ts timestamptz=now(), start_block numeric=0, end_block numeric=9e18) RETURNS integer
+LANGUAGE plpgsql AS $function$
+DECLARE r integer;
+BEGIN
+WITH rows AS (
+    INSERT INTO aave.tokens (
+    symbol,
+    contract_address,
```
Comment (chuxinh) [inline optimism2/aave/tokens.sql:19]: `token` or `token_address`?

File: optimism2/aave/interest.sql [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    symbol,
+    reserve,
+    deposit_apy,
+    stable_borrow_apy,
+    variable_borrow_apy,
+    deposit_apr AS daily_deposit_apr,
+    stable_borrow_apr AS daily_stable_borrow_apr,
+    variable_borrow_apr AS daily_variable_borrow_apr
+FROM (
+SELECT
```
Comment (chuxinh) [inline optimism2/aave/interest.sql:43]: indent below

File: polygon/aave/tokens.sql [hunk 6]
```diff
@@ -0,0 +1,79 @@
+CREATE TABLE IF NOT EXISTS aave.tokens (   
+    version text,
+    symbol  text,
+    contract_address bytea,
+    "aToken" bytea,
+    "stableDebtToken" bytea,
+    "variableDebtToken" bytea,
```
Comment (chuxinh) [inline polygon/aave/tokens.sql:7]: update to `lower_snake_cased`

File: polygon/aave/borrow.sql [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+      liquidator,
+      amount / (10^erc20.decimals) AS amount,
+      (amount/(10^p.decimals)) * price AS usd_amount,
+      evt_tx_hash,
+      evt_index,
+      evt_block_time,
+      evt_block_number
+    
+FROM (
+-- v2
```
Comment (chuxinh) [inline polygon/aave/borrow.sql:58]: indent below

File: polygon/aave/borrow.sql [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+SELECT 
+    '2' AS version,
+    'borrow' AS transaction_type,
+    CASE 
+    WHEN "borrowRateMode" = '1' THEN 'stable'
+    WHEN "borrowRateMode" = '2' THEN 'variable'
```
Comment (chuxinh) [inline polygon/aave/borrow.sql:64]: indent

File: polygon/aave/borrow.sql [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    evt_block_number
+FROM aave_v2."LendingPool_evt_LiquidationCall"
+UNION ALL 
+-- v3
+SELECT 
+    '3' AS version,
+    'borrow' AS transaction_type,
+    CASE 
+    WHEN "interestRateMode" = '1' THEN 'stable'
+    WHEN "interestRateMode" = '2' THEN 'variable'
```
Comment (chuxinh) [inline polygon/aave/borrow.sql:115]: indent

File: polygon/aave/borrow.sql [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+    - amount AS amount,
+    evt_tx_hash,
+    evt_index,
+    evt_block_time,
+    evt_block_number
+FROM aave_v3."Pool_evt_Repay"
+UNION ALL 
+-- liquidation
+SELECT 
+    '2' AS version,
```
Comment (chuxinh) [inline polygon/aave/borrow.sql:146]:
```suggestion
    '3' AS version,
```

File: polygon/aave/deposit.sql [hunk 11]
```diff
@@ -0,0 +1,180 @@
+CREATE TABLE IF NOT EXISTS aave.deposit (   
+    version text,
+    transaction_type text,
+    symbol text,
+    contract_address bytea,
+    borrower bytea,
+    "to" bytea,
```
Comment (chuxinh) [inline polygon/aave/deposit.sql:7]: rename accordingly

File: polygon/aave/deposit.sql [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+      liquidator,
+      amount / (10^erc20.decimals) AS amount,
+      (amount/(10^p.decimals)) * price AS usd_amount,
+      evt_tx_hash,
+      evt_index,
+      evt_block_time,
+      evt_block_number
+    
+FROM (
+
```
Comment (chuxinh) [inline polygon/aave/deposit.sql:55]: indent below

File: optimism2/aave/interest.sql [hunk 13]
```diff
@@ -0,0 +1,113 @@
+CREATE TABLE IF NOT EXISTS aave.interest (   
+    day timestamptz,
```
Comment (chuxinh) [inline optimism2/aave/interest.sql:2]: might be worth having a version here as well despite optimism only having v3

File: ethereum/aave/interest.sql [hunk 14]
```diff
@@ -0,0 +1,112 @@
+CREATE TABLE IF NOT EXISTS aave.interest (   
+    day timestamptz,
```
Comment (chuxinh) [inline ethereum/aave/interest.sql:2]: same that it might be worth having a version here too

**Review by chuxinh — CHANGES_REQUESTED** [review]
Thanks for the great work @Treaqura!
Have some nits here and there but think the overall logic looks good to me. Some of the feedback is suggestive so let me know if you think otherwise!

Make sure while you're reflecting changes, it applies to all the other protocols :D

File: ethereum/aave/borrow.sql [hunk 15]
```diff
@@ -0,0 +1,201 @@
...
+SELECT 
+    '2' AS version,
+    'borrow_liquidation' AS transaction_type,
+    NULL AS loan_type,
+    CASE
+    WHEN _reserve = '\xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE' THEN '\xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2' 
+    ELSE _reserve
+    END AS contract_address,
+    "_user" AS borrower,
+    _liquidator AS repayer,
```
[the remark this answers is not in the document]
Comment (Treaqura) [reply ethereum/aave/borrow.sql:113]: I think if a user gets liquidated then technically the loan gets repayed by the user that’s liquidating the position, hence why I’ve listed the liquidator as both the repayer and the liquidator. I could list the original user as the repayer but I think that’s confusing and not really correct. What do you think?

File: ethereum/aave/interest.sql [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+    ((1+stable_borrow_apy)^(1.0/365.0)-1) AS stable_borrow_apr,
+    ((1+variable_borrow_apy)^(1.0/365.0)-1) AS variable_borrow_apr
+FROM (
+SELECT 
+    date_trunc('day', evt_block_time) AS day,
+    reserve,
+    AVG("liquidityRate" / 1e27) AS deposit_apy,
+    AVG("stableBorrowRate" / 1e27) AS stable_borrow_apy,
+    AVG("variableBorrowRate" /1e27) AS variable_borrow_apy
+FROM aave_v2."LendingPool_evt_ReserveDataUpdated"
```
[the remark this answers is not in the document]
Comment (Treaqura) [reply ethereum/aave/interest.sql:73]: Yes, this table is not in v1 unfortunately

File: optimism2/aave/interest.sql (hunk 13 above)
Comment (Treaqura) [reply optimism2/aave/interest.sql:3]: Agree, I've added this for Optimism aswell

**Review by Treaqura — COMMENTED** [review]
Thanks for the review @chuxinh ! I've updated all tables with your feedback. 

For the deposit tables I've changed borrower to depositor for it to be consistent with the borrow.sql tables. "to" has been changed to withdrawn_to as this is the wallet to deposit was withdrawn to and should be more clear.

File: optimism2/aave/tokens.sql [hunk 17]
```diff
@@ -0,0 +1,67 @@
+CREATE TABLE IF NOT EXISTS aave.tokens (   
+    symbol text,
```
Comment (chuxinh) [inline optimism2/aave/tokens.sql:2]: just for consistency, add version here as well for optimism (similar to that for polygon)

File: ethereum/aave/tokens.sql [hunk 18]
```diff
@@ -0,0 +1,69 @@
+CREATE TABLE IF NOT EXISTS aave.tokens (
```
Comment (chuxinh) [inline ethereum/aave/tokens.sql:1]: add version for this one too

**Review by chuxinh — COMMENTED** [review]
a few more nits and LGTM after that! 👏 

could you also provide some tests query on these abstractions and see if the aggregated value can add up to the amount borrowed and available and compare that with what AAVE has? 

also adding Dune team here to take a look @masquot @jeff-dude 🙏

**Comment by jeff-dude** [comment]
thank you @Treaqura @chuxinh for working through this. i'm happy to look into it tomorrow, if it's ready

**Pull request merged**